src/symbol/cross_tile_symbol_index.js

```diff
@@ -85,60 +82,75 @@ class TileLayerIndex {
-            // Return any symbol with the same keys whose coordinates are within 1
-            // grid unit. (with a 4px grid, this covers a 12px by 12px area)
-            if (Math.abs(thisTileSymbol.coordinates.x - childTileSymbolCoordinates.x) <= 1 &&
-                Math.abs(thisTileSymbol.coordinates.y - childTileSymbolCoordinates.y) <= 1) {
-                return thisTileSymbol.instance;
+            const indexedInstances = this.indexedSymbolInstances[symbolInstance.key];
+            if (!indexedInstances) {
+                // No symbol with this key in this bucket
+                continue;
             }
-        }
-    }
 
-    forEachSymbolInstance(fn: (SymbolInstance) => void) {
-        for (const key in this.symbolInstances) {
-            const keyedSymbolInstances = this.symbolInstances[key];
-            for (const symbolInstance of keyedSymbolInstances) {
-                fn(symbolInstance.instance);
+            const scaledSymbolCoord = this.getScaledCoordinates(symbolInstance, newTileID);
+
+            for (const thisTileSymbol of indexedInstances) {
+                // Return any symbol with the same keys whose coordinates are within 1
+                // grid unit. (with a 4px grid, this covers a 12px by 12px area)
+                if (Math.abs(thisTileSymbol.coord.x - scaledSymbolCoord.x) <= tolerance &&
+                    Math.abs(thisTileSymbol.coord.y - scaledSymbolCoord.y) <= tolerance) {
+                    symbolInstance.crossTileID = thisTileSymbol.crossTileID;
+                    break;
+                }
             }
         }
     }
 }
 
+class CrossTileIDs {
+    maxCrossTileID: number;
+    constructor() {
+        this.maxCrossTileID = 0;
+    }
+    generate() {
+        return ++this.maxCrossTileID;
+    }
+}
+
 class CrossTileSymbolLayerIndex {
-    indexes: {[zoom: number]: {[tileId: number]: TileLayerIndex}};
+    indexes: {[zoom: string | number]: {[tileId: string | number]: TileLayerIndex}};
 
     constructor() {
         this.indexes = {};
     }
 
-    addTile(tileID: OverscaledTileID, symbolInstances: Array<SymbolInstance>) {
+    addBucket(tileID: OverscaledTileID, bucket: SymbolBucket, crossTileIDs: CrossTileIDs) {
+        if (this.indexes[tileID.overscaledZ] &&
+            this.indexes[tileID.overscaledZ][tileID.key] &&
+            this.indexes[tileID.overscaledZ][tileID.key].bucketInstanceId === bucket.bucketInstanceId) {
+            return false;
+        }
 
         let minZoom = 25;
         let maxZoom = 0;
         for (const zoom in this.indexes) {
-            minZoom = Math.min((zoom: any), minZoom);
-            maxZoom = Math.max((zoom: any), maxZoom);
+            minZoom = Math.min(+zoom, minZoom);
+            maxZoom = Math.max(+zoom, maxZoom);
         }
 
-        const tileIndex = new TileLayerIndex(tileID, symbolInstances);
-
         // make all higher-res child tiles block duplicate labels in this tile
```

Comment:
These two loops (child index and parent index) still have structure and comments from when child and parent indexes had different blocking logic. Now that the logic is really just "run `findMatches` across all zoom levels", maybe we could express this more clearly with one loop that just has a branch for looking up child vs. parent indices?